### PR TITLE
[FLINK-17118][python] Add Cython support for primitive data types

### PR DESCRIPTION
**Pull request by HuangXingBo**
## What is the purpose of the change

*This pull request will support primitive DataTypes in Cython*

## Brief change log

  - *Adds the file fast_coder_impl.(pyx,pxd) which includes the implementation of cython*
  - *Add cython test environment in tox*

## Verifying this change
  - *Add correponding unit test in test_coders_common.py*
  - *Add cython test environment in tox*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-python/pyflink/fn_execution/fast_coder_impl.pyx [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+        self.encode_row_result(wrapper_stream, out_stream)
+
+    cdef encode_row_result(self, WrapperFuncInputElement wrapper_func_input_element,
+                           OutputStream out_stream):
+        cdef list result
+        self._before_encode(wrapper_func_input_element, out_stream)
+        while self._input_buffer_size > self._input_pos:
+            self._load_row()
+            result = self.func(self.row)
+            self._write_data(result)
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/fast_coder_impl.pyx:93]: _write_data -> _encode_one_row

File: flink-python/pyflink/fn_execution/fast_coder_impl.pyx [hunk 2]
```diff
@@ -0,0 +1,558 @@
...
+            return self._load_double()
+        elif field_type == BINARY:
+            # bytes
+            return self._load_bytes()
+        elif field_type == CHAR:
+            # str
+            return self._load_bytes().decode("utf-8")
+        elif field_type == DATE:
+            # Date
+            return datetime.date.fromordinal(self._load_int() + 719163)
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/fast_coder_impl.pyx:217]: Could you add some explain how 719163 is computed?

File: flink-python/pyflink/fn_execution/fast_coder_impl.pyx [hunk 3]
```diff
@@ -0,0 +1,558 @@
...
+        self._null_byte_search_table[3] = 0x10
+        self._null_byte_search_table[4] = 0x08
+        self._null_byte_search_table[5] = 0x04
+        self._null_byte_search_table[6] = 0x02
+        self._null_byte_search_table[7] = 0x01
+        for i in range(self._output_field_count):
+            self._output_field_type[i] = self._output_field_coders[i].type_name()
+            self._output_coder_type[i] = self._output_field_coders[i].coder_type()
+
+    cdef void _consume_input_data(self, WrapperInputElement wrapper_input_element, size_t size):
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/fast_coder_impl.pyx:125]: _wrapInputStream

File: flink-python/pyflink/fn_execution/fast_coder_impl.pyx [hunk 4]
```diff
@@ -0,0 +1,558 @@
...
+        self._output_row_data[self._output_row_pos] = val
+        self._output_row_pos += 1
+
+    cdef void _dump_smallint(self, libc.stdint.int16_t v):
+        if self._output_row_buffer_size < self._output_row_pos + 2:
+            self._output_row_buffer_size *= 2
+            self._output_row_data = <char*> libc.stdlib.realloc(self._output_row_data,
+                                                                self._output_row_buffer_size)
+        self._output_row_data[self._output_row_pos] = <unsigned char> (v >> 8)
+        self._output_row_data[self._output_row_pos + 1] = <unsigned char> (v)
```
Comment (dianfu) [inline flink-python/pyflink/fn_execution/fast_coder_impl.pyx:399]: Remove the parentheses

**Comment by HuangXingBo** [comment]
Thanks a lot for @dianfu review, I have addressed the comments at the latest commit.

**Comment by dianfu** [comment]
@HuangXingBo Thanks a lot for the update. The test time increases a lot, could you take a look?

**Comment by dianfu** [comment]
It seems that it's because we run the udf related tests also in cython cases.

**Pull request merged**